src/transactions.rs

```diff
@@ -0,0 +1,456 @@
+    pub fn with_timeout(&self, timeout: u64) -> Result<Vec<Transaction<L>>, KeystoreError<L>> {
+        let uids = self
+            .expiring_txns
+            .get(&timeout)
+            .ok_or(KeyValueStoreError::KeyNotFound)?;
+
+        Ok(uids.iter().map(|uid| self.get(uid).unwrap()).collect())
+    }
+
+    /// Commit the store version.
+    pub fn commit(&mut self) -> Result<(), KeystoreError<L>> {
+        Ok(self.store.commit_version()?)
+    }
+
+    /// Revert the store version.
+    pub fn revert(&mut self) -> Result<(), KeystoreError<L>> {
+        Ok(self.store.revert_version()?)
+    }
+
+    /// Add a transaction has to the index and update the stored transaction with this has
+    pub fn insert_hash(
+        &mut self,
+        hash: TransactionHash<L>,
+        uid: &TransactionUID<L>,
+    ) -> Result<(), KeystoreError<L>> {
+        let editor = self.get_mut(uid)?;
+        editor.with_hash(hash.clone()).save()?;
+        self.txn_by_hash.insert(hash, uid.clone());
+        Ok(())
+    }
+
+    /// Add a list of Memo ids a transaction is waiting for.  We update the index and the stored transaction
+    pub fn add_pending_uids(
+        &mut self,
+        txn_uid: &TransactionUID<L>,
+        pending_uids: Vec<u64>,
+    ) -> Result<(), KeystoreError<L>> {
+        let txn_editor = self.get_mut(txn_uid)?;
+        txn_editor.add_pending_uids(&pending_uids).save()?;
+        for uid in pending_uids {
+            self.uids_awaiting_memos.insert(uid, txn_uid.clone());
+        }
+
+        Ok(())
+    }
+
+    /// Remove pending memo ids from the index and the stored transactions which were awaiting those ids
+    pub async fn remove_pending_uids(
+        &mut self,
+        pending_uids: Vec<u64>,
+    ) -> Result<(), KeystoreError<L>> {
+        for uid in pending_uids {
+            if let Some(txn_uid) = self.uids_awaiting_memos.remove(&uid) {
+                let txn_editor = self.get_mut(&txn_uid)?;
+                txn_editor.remove_pending_uid(uid).save()?;
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Remove a transaction from the pending index when it
+    pub async fn remove_pending(
+        &mut self,
+        timeout: u64,
+        uid: &TransactionUID<L>,
+    ) -> Result<(), KeystoreError<L>> {
+        if let Some(expiring) = self.expiring_txns.get_mut(&timeout) {
+            expiring.remove(uid);
+            if expiring.is_empty() {
+                self.expiring_txns.remove(&timeout);
+            }
+        }
+        Ok(())
+    }
+
+    /// Create an Transaction.
+    ///
+    /// When we first create a transaction it won't know all of it's values until
+    /// specific events happen on chain.  When the block for this transaction commited
+    /// We will learn it's Hash, Recept, and PendingUIDs as well as be certain of the
+    /// Fee and Asset Change records.
+    ///
+    /// Returns the editor for the created asset.
+    #[allow(clippy::too_many_arguments)]
+    pub fn create(
```

Comment:
With so many parameters, it may be worth creating a `TransactionParams` struct, which at least lets the caller specify the parameters by name (e.g. `TransactionParams { uid: something, timeout: t, ... }`)